rust/core-lib/src/event_context.rs

```diff
@@ -494,6 +494,32 @@ impl<'a> EventContext<'a> {
         self.editor.borrow_mut().dec_revs_in_flight();
     }
 
+    /// Returns the text to be saved, appending a newline if necessary.
+    pub(crate) fn text_for_save(&mut self) -> Rope {
+        let editor = self.editor.borrow();
+        let mut rope = editor.get_buffer().clone();
+        let rope_len = rope.len();
+
+        if rope_len < 1 {
+            return rope;
+        }
+
+        let cursor = Cursor::new(&rope, rope.len());
+        let has_newline_at_eof = match cursor.get_leaf() {
+            Some((last_chunk, _)) => last_chunk.ends_with(&self.config.line_ending),
+            // The rope can't be empty, since we would have returned earlier if it was
+            None => unreachable!(),
```

Comment:
I guess we could just return rope here to handle this gracefully.

Reply:
If we ever reach this point something is fundamentally wrong with our assumptions, and we probably prefer to crash, I think?

This annotation is also a hint to the compiler; knowing that some point can't be reached can allow for additional optimizations.

Reply:
Ah, alright.